Tasks/ANT/anttask.ts

```diff
@@ -125,80 +40,232 @@ function publishTestResults(publishJUnitResults, testResultsFiles: string) {
+                // jdkVersion should be in the form of 1.7, 1.8, or 1.10
+                // jdkArchitecture is either x64 or x86
+                // envName for version 1.7 and x64 would be "JAVA_HOME_7_X64"
+                var envName = "JAVA_HOME_" + jdkVersion.slice(2) + "_" + jdkArchitecture.toUpperCase();
+                specifiedJavaHome = tl.getVariable(envName);
+                if (!specifiedJavaHome) {
+                    throw new Error('Failed to find specified JDK version. Please make sure environment variable ' + envName + ' exists and is set to the location of a corresponding JDK.');
+                }
+            }
+        }
+        else {
+            tl.debug('Using path from user input to set JAVA_HOME');
+            var jdkUserInputPath = tl.getPathInput('jdkUserInputPath', true, true);
+            specifiedJavaHome = jdkUserInputPath;
+        }
+
+        if (specifiedJavaHome) {
+            tl.debug('Set JAVA_HOME to ' + specifiedJavaHome);
+            process.env['JAVA_HOME'] = specifiedJavaHome;
+        }
+
+        var ccTool = tl.getInput('codeCoverageTool');
+        var isCodeCoverageOpted = (typeof ccTool != "undefined" && ccTool && ccTool.toLowerCase() != 'none');
+
+        var buildRootPath = path.dirname(antBuildFile);
+        var instrumentedClassesDirectory = path.join(buildRootPath, "InstrumentedClasses");
+        //delete any previous cobertura instrumented classes as they might interfere with ant execution.
+        tl.rmRF(instrumentedClassesDirectory, true);
+
+        if (isCodeCoverageOpted) {
+            var summaryFile: string = null;
+            var reportDirectory: string = null;
+            var ccReportTask: string = null;
+            var reportBuildFile: string = null;
+            enableCodeCoverage();
+        }
+        else {
+            tl.debug("Option to enable code coverage was not selected and is being skipped.");
+        }
+
+        var publishJUnitResults = tl.getInput('publishJUnitResults');
+        var testResultsFiles = tl.getInput('testResultsFiles', true);
+
+        await antv.exec();
+
+        var buffer;
+        antb.on('stdout', (data) => {
+            if (data) {
+                buffer += data.toString();
+
+                let idx = buffer.indexOf(os.EOL);
+                while (idx > -1) {
+                    let line = buffer.substring(0, idx);
+
+                    processAntOutputLine(line);
+
+                    buffer = buffer.substring(idx + os.EOL.length);
+                    idx = buffer.indexOf(os.EOL);
+                }
             }
-        }).fail(function(err) {
-            tl.warning("No code coverage results found to be published. This could occur if there were no tests executed or there was a build failure. Check the ant output for details.");
         });
+
+        antb.exec()
+            .then(function (code) {
+                publishTestResults(publishJUnitResults, testResultsFiles);
```

Comment:
If you use await, you could publishTestResults and publishCodeCoverage inside finally block. This is fine too, just pointing out an alternate.